pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+    # __isub__ uses NumericValue base class implementation
+    # __imul__ uses NumericValue base class implementation
+    # __idiv__ uses NumericValue base class implementation
+    # __itruediv__ uses NumericValue base class implementation
+    # __ipow__ uses NumericValue base class implementation
+    # __neg__ uses NumericValue base class implementation
+    # __pos__ uses NumericValue base class implementation
+    # __add__ uses NumericValue base class implementation
+
+    def __str__(self):
+        """ Returns a string representing the unit """
+
+        # The ~ returns the short form of the pint unit if the unit is
+        # an instance of the unit 'dimensionless', then pint returns ''
+        # which causes problems with some string processing in Pyomo
+        # that expects a name
+        #
+        # Note: Some pint units contain unicode characters (notably
+        # delta temperatures).  So that things work cleanly in Python 2
+        # and 3, we will generate the string as unicode, then explicitly
+        # encode it to UTF-8 in Python 2
+        retstr = u'{:!~s}'.format(self._pint_unit)
+        if retstr == '':
+            retstr = 'dimensionless'
+        if six.PY2:
+            return str(retstr.encode('utf8'))
+        else:
+            return retstr
+
+    def to_string(self, verbose=None, labeler=None, smap=None,
+                  compute_values=False):
+        """
+        Return a string representation of the expression tree.
+
+        See documentation on :py:class:`NumericValue`
+
+        Returns
+        -------
+        : bool
+           A string representation for the expression tree.
+        """
+        return str(self)
+
+    def __nonzero__(self):
+        """Unit is treated as a constant value of 1.0. Therefore, it is always nonzero
+        Returns
+        -------
+        : bool
+           Returns whether on not the object is non-zero
+        """
+        return self.__bool__()
+
+    def __bool__(self):
+        """Unit is treated as a constant value of 1.0. Therefore, it is always "True"
+
+        Returns
+        -------
+        : bool
+           Returns whether or not the object is "empty"
+        """
+        return True
+
+    def __call__(self, exception=True):
+        """Unit is treated as a constant value, and this method always returns 1.0
+
+        Returns
+        -------
+        : float
+           Returns 1.0
+        """
+        return 1.0
+
+    def pprint(self, ostream=None, verbose=False):
+        """Display a user readable string description of this object.
+        """
+        if ostream is None: #pragma:nocover
+            ostream = sys.stdout
+        ostream.write(str(self))
+        # There is also a long form, but the verbose flag is not really the correct indicator
+        # if verbose:
+        #     ostream.write('{:!s}'.format(self._pint_unit))
+        # else:
+        #     ostream.write('{:!~s}'.format(self._pint_unit))
+
+def _is_leaf(expr):
```

Comment:
This is a trivial function actively used in only one place in the code (and that location is inside a tight loop).  [It also appears in an assertion].  Consider in-lining.

Reply:
I purposely did not inline this because it was not at all obvious to me when first looking at the expression code that those checks were, in fact, checking for a leaf. I will change it as you indicate, however, we might want to work on the naming to make this more clear.